File: app/src/main/java/com/commonpepper/photosen/ui/activities/SearchActivity.java
package com.commonpepper.photosen.ui.activities;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.KeyEvent;
import android.view.View;
import android.view.WindowManager;
import android.view.inputmethod.EditorInfo;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;

import com.commonpepper.photosen.R;
import com.commonpepper.photosen.ui.fragments.SearchListFragment;
import com.commonpepper.photosen.ui.viewmodels.SearchActivityViewModel;
import com.google.android.material.chip.Chip;
import com.google.android.material.chip.ChipGroup;

import java.util.HashSet;
import java.util.Set;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.lifecycle.ViewModelProviders;

public class SearchActivity extends AppCompatActivity {
    public static final String TAG_SEARCHTAG = "search_tag";
    private static final String TAG_RECREATED = "rotated_tag";

    private EditText searchText;
    private ChipGroup chipGroup;
    private Chip firstChip;
    private EditText inputTag;

    private Set<String> tags = new HashSet<>();
    private SearchActivityViewModel viewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_search);

        searchText = findViewById(R.id.search_edit_text);
        Toolbar toolbar = findViewById(R.id.search_toolbar);
        chipGroup = findViewById(R.id.search_chip_group);
        firstChip = findViewById(R.id.search_first_chip);
        inputTag = findViewById(R.id.tag_input);

        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setDisplayShowHomeEnabled(true);

        Intent intent = getIntent();
        String firstTag = intent.getStringExtra(TAG_SEARCHTAG);
        boolean recreated = false;
        if (savedInstanceState != null) recreated = savedInstanceState.getBoolean(TAG_RECREATED);
        viewModel = ViewModelProviders.of(this).get(SearchActivityViewModel.class);
        if (viewModel.getTags().size() > 0 || firstTag == null || recreated) {
            for (String tag : viewModel.getTags()) {
                addNewChip(tag);
            }
            chipGroup.removeView(firstChip);
            if (!recreated) getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_VISIBLE);
        } else {
            viewModel.getTags().add(firstTag);
            firstChip.setText(firstTag);
            firstChip.setOnCloseIconClickListener(onChipClickListener);
            doSearch();
        }

        searchText.setOnEditorActionListener((v, actionId, event) -> {
            doSearch();
            return true;
        });

        inputTag.setOnEditorActionListener((v, actionId, event) -> {
            if (actionId == EditorInfo.IME_ACTION_SEARCH || actionId == EditorInfo.IME_ACTION_DONE ||
                    event != null && event.getAction() == KeyEvent.ACTION_DOWN
                            && event.getKeyCode() == KeyEvent.KEYCODE_ENTER) {
                if (event == null || !event.isShiftPressed()) {
                    String newTag = inputTag.getText().toString();
                    if (newTag.length() > 0 && !tags.contains(newTag)) {
                        addNewChip(newTag);
                        viewModel.getTags().add(newTag);
                        inputTag.setText("");
                        doSearch();
                    }
                    return true;
                }
            }
            return false;
        });
    }

    private void addNewChip(String newTag) {
        Chip chip = new Chip(this);
        chip.setText(newTag);
        chip.setLayoutDirection(View.LAYOUT_DIRECTION_LOCALE);
        chip.setCloseIconVisible(true);
        chip.setOnCloseIconClickListener(onChipClickListener);
        chipGroup.addView(chip);
    }

    private View.OnClickListener onChipClickListener = new View.OnClickListener() {
        @Override
        public void onClick(View x) {
            Chip chip = (Chip) x;
            chipGroup.removeView(chip);
            viewModel.getTags().remove(chip.getText().toString());
            SearchActivity.this.doSearch();
        }
    };

    @Override
    public boolean onSupportNavigateUp() {
        onBackPressed();
        return true;
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (searchText.getText().toString().length() > 0) {
            getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_HIDDEN);
        }
    }

    private void doSearch() {
        String query = searchText.getText().toString();
        if (query.length() == 0) query = null;

        String tagsExtra = viewModel.tagsToString();
        if (query != null || tagsExtra != null) {
            getSupportFragmentManager().beginTransaction()
                    .replace(R.id.search_linear_layout, SearchListFragment.newInstance(query, tagsExtra))
                    .commit();

            View view = this.getCurrentFocus();
            if (view == null) view = new View(this);
            InputMethodManager imm = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
            imm.hideSoftInputFromWindow(view.getWindowToken(), 0);
        }
    }

    @Override
    protected void onSaveInstanceState(Bundle savedInstanceState) {
        savedInstanceState.putBoolean(TAG_RECREATED, true);
        super.onSaveInstanceState(savedInstanceState);
    }

}

File: app/src/main/java/com/commonpepper/photosen/ui/activities/SearchActivity.kt
package com.commonpepper.photosen.ui.activities

import android.content.Context
import android.os.Bundle
import android.view.KeyEvent
import android.view.View
import android.view.View.OnClickListener
import android.view.WindowManager.LayoutParams
import android.view.inputmethod.EditorInfo
import android.view.inputmethod.InputMethodManager
import android.widget.EditText
import androidx.drawerlayout.widget.DrawerLayout
import androidx.lifecycle.ViewModelProviders
import com.commonpepper.photosen.R.*
import com.commonpepper.photosen.ui.adapters.MyPagerAdapter
import com.commonpepper.photosen.ui.fragments.SearchListFragment.Companion.newInstance
import com.commonpepper.photosen.ui.viewmodels.SearchActivityViewModel
import com.google.android.material.chip.Chip
import com.google.android.material.chip.ChipGroup
import kotlinx.android.synthetic.main.activity_search.*
import kotlinx.android.synthetic.main.navigation_view.*

class SearchActivity : AbstractNavActivity() {
    override val abstractDrawerLayout: DrawerLayout get() = drawerLayout
    private lateinit var pagerAdapter: MyPagerAdapter
    private lateinit var viewModel: SearchActivityViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(layout.activity_search)
        setSupportActionBar(toolbar)
        supportActionBar?.apply {
            setDisplayHomeAsUpEnabled(true)
            setDisplayShowHomeEnabled(true)
        }
        navigationView.menu.findItem(id.drawer_search).apply {
            isCheckable = true
            isChecked = true
        }
        navigationView.setNavigationItemSelectedListener(this)
        pagerAdapter = MyPagerAdapter(supportFragmentManager)
        viewPager.adapter = pagerAdapter
        tabLayout.setupWithViewPager(viewPager)
        val firstTag: String? = intent.getStringExtra(TAG_SEARCHTAG)
        val recreated = savedInstanceState?.getBoolean(TAG_RECREATED) ?: false
        viewModel = ViewModelProviders.of(this)[SearchActivityViewModel::class.java]
        if (viewModel.tags.size > 0 || firstTag == null || recreated) {
            viewModel.tags.forEach { addNewChip(it) }
            searchChipGroup.removeView(searchFirstChip)
            if (!recreated) {
                window.setSoftInputMode(LayoutParams.SOFT_INPUT_STATE_VISIBLE)
            } else {
                doSearch()
            }
        } else {
            viewModel.tags.add(firstTag)
            searchFirstChip.text = firstTag
            searchFirstChip.setOnCloseIconClickListener(onChipClickListener)
            doSearch()
        }
        searchText.setOnEditorActionListener { _, _, _ ->
            viewModel.query = searchText!!.text.toString()
            doSearch()
            true
        }
        tagInput.setOnEditorActionListener { _, actionId: Int, event: KeyEvent? ->
            if (actionId == EditorInfo.IME_ACTION_SEARCH || actionId == EditorInfo.IME_ACTION_DONE || event != null && event.action == KeyEvent.ACTION_DOWN && event.keyCode == KeyEvent.KEYCODE_ENTER) {
                if (event == null || !event.isShiftPressed) {
                    val newTag = tagInput!!.text.toString()
                    if (newTag.isNotEmpty() && !viewModel.tags.contains(newTag)) {
                        addNewChip(newTag)
                        viewModel.tags.add(newTag)
                        tagInput.setText("")
                        doSearch()
                    }
                    return@setOnEditorActionListener true
                }
            }
            false
        }
    }

    private fun addNewChip(newTag: String) {
        val chip = Chip(this)
        chip.text = newTag
        chip.layoutDirection = View.LAYOUT_DIRECTION_LOCALE
        chip.isCloseIconVisible = true
        chip.setOnCloseIconClickListener(onChipClickListener)
        searchChipGroup.addView(chip)
    }

    private val onChipClickListener = OnClickListener { x ->
        val chip = x as Chip
        searchChipGroup.removeView(chip)
        viewModel.tags.remove(chip.text.toString())
        doSearch()
    }

    override fun onSupportNavigateUp(): Boolean {
        onBackPressed()
        return true
    }

    override fun onResume() {
        super.onResume()
        if (searchText.text.toString().isNotEmpty()) {
            window.setSoftInputMode(LayoutParams.SOFT_INPUT_STATE_HIDDEN)
        }
    }

    private fun doSearch() {
        val query = viewModel.query
        val tagsExtra = viewModel.tags.joinToString(",")
        if ((query != null && query.isNotEmpty()) || tagsExtra.isNotEmpty()) {
            tabLayout.visibility = View.VISIBLE
            val fragmentRelevant = newInstance(query, tagsExtra, "relevance")
            val fragmentMostViewed = newInstance(query, tagsExtra, "interestingness-desc")
            val fragmentLatest = newInstance(query, tagsExtra, "date-posted-desc")
            pagerAdapter.apply {
                clear()
                addFragment(fragmentRelevant, getString(string.relevant))
                addFragment(fragmentMostViewed, getString(string.most_viewed))
                addFragment(fragmentLatest, getString(string.latest))
            }
            val view = this.currentFocus ?: View(this)
            val imm = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
            imm.hideSoftInputFromWindow(view.windowToken, 0)
        }
    }

    override fun onSaveInstanceState(savedInstanceState: Bundle) {
        savedInstanceState.putBoolean(TAG_RECREATED, true)
        super.onSaveInstanceState(savedInstanceState)
    }

    companion object {
        const val TAG_SEARCHTAG = "search_tag"
        const val TAG_RECREATED = "rotated_tag"
    }
}
